playground/engine.py

```diff
@@ -0,0 +1,818 @@
+            refractive_index_tensor=torch.cat([p.refractive_index_tensor for p in primitives], dim=0).float()
+        )
+
+    def apply_transform(self):
+        model_matrix = self.transform.model_matrix()
+        rs_comp = model_matrix[None, :3, :3]
+        t_comp = model_matrix[None, :3, 3:]
+        transformed_verts = (rs_comp @ self.vertices[:, :, None] + t_comp).squeeze(2)
+
+        normal_matrix = self.transform.rotation_matrix()[None, :3, :3]
+        transformed_normals = (normal_matrix @ self.vertex_normals[:, :, None]).squeeze(2)
+        transformed_normals = torch.nn.functional.normalize(transformed_normals)
+
+        transformed_tangents = (normal_matrix @ self.vertex_tangents[:, :, None]).squeeze(2)
+        transformed_tangents = torch.nn.functional.normalize(transformed_tangents)
+
+        return OptixPrimitive(
+            vertices=transformed_verts,
+            triangles=self.triangles,
+            vertex_normals=transformed_normals,
+            vertex_tangents=transformed_tangents,
+            has_tangents=self.has_tangents,
+            material_uv=self.material_uv,
+            material_id=self.material_id,
+            primitive_type=self.primitive_type,
+            primitive_type_tensor=self.primitive_type_tensor,
+            reflectance_scatter=self.reflectance_scatter,
+            refractive_index=self.refractive_index,
+            refractive_index_tensor=self.refractive_index_tensor,
+            transform=ObjectTransform(device=self.transform.device)
+        )
+
+
+class Primitives:
+    SUPPORTED_MESH_EXTENSIONS = ['.obj', '.glb']
+    DEFAULT_REFRACTIVE_INDEX = 1.33
+    SCALE_OF_NEW_MESH_TO_SMALL_SCENE = 0.5    # Mesh will be this percent of the scene on longest axis
+
+    def __init__(self, tracer, mesh_assets_folder, enable_envmap=False, use_envmap_as_background=False,
+                 scene_scale=None):
+        # str -> str ; shape name to filename + extension
+        self.assets = self.register_available_assets(assets_folder=mesh_assets_folder)
+        self.tracer = tracer
+        self.enabled = True
+        self.objects = dict()
+        self.use_smooth_normals = True
+        self.disable_gaussian_tracing = False
+        self.disable_pbr_textures = False
+        self.force_white_bg = False
+        self.enable_envmap = enable_envmap
+        self.use_envmap_as_background = use_envmap_as_background
+
+        if scene_scale is None:
+            self.scene_scale = torch.tensor([1.0, 1.0, 1.0], device='cpu')
+        else:
+            self.scene_scale = scene_scale.cpu()
+
+        self.stacked_fields = None
+        self.dirty = True
+
+        self.instance_counter = dict()  # Counts number of primitives of each geometry
+
+        device = 'cuda'
+
+        self.registered_materials = self.register_default_materials(device)  # str -> PBRMaterial
+
+    def register_available_assets(self, assets_folder):
```

Comment:
here and below, if you intend Primitives class to be used by library clients, please add docstrings and sample usage.